fraud/service.go

```diff
@@ -0,0 +1,136 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// service is responsible for validating and propagating Fraud Proofs.
+// It implements the Service interface.
+type service struct {
+	mu           sync.RWMutex
+	unmarshalers map[ProofType]ProofUnmarshaler
+
+	topics *topics
+	getter headerFetcher
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.Lock()
+	u, ok := f.unmarshalers[proofType]
+	if !ok {
+		return nil, errors.New("fraud: unmarshaler is not registered")
+	}
+
+	t, ok := f.topics.getTopic(proofType)
+
+	// if topic was not stored in cache, then we should register a validator and
+	// join pubsub topic.
+	if !ok {
+		err := f.topics.registerValidator(proofType, f.processIncoming)
```

Comment:
Better to register validator after joining

Reply:
done